src/main/scala/is/hail/annotations/MemoryBlock.scala

```diff
@@ -225,9 +225,158 @@ final class MemoryBuffer(var mem: Long, var length: Long, var offset: Long = 0)
   override def finalize() {
     Memory.free(mem)
   }
+
+  def visit(t: Type, off: Long, v: ValueVisitor) {
+    t match {
+      case TBoolean => v.visitBoolean(loadBoolean(off))
+      case TInt32 => v.visitInt32(loadInt(off))
+      case TInt64 => v.visitInt64(loadLong(off))
+      case TFloat32 => v.visitFloat32(loadFloat(off))
+      case TFloat64 => v.visitFloat64(loadDouble(off))
+      case TString =>
+        val boff = loadAddress(off)
+        v.visitString(TString.loadString(this, boff))
+      case TBinary =>
+        val boff = loadAddress(off)
+        val length = TBinary.loadLength(this, boff)
+        val b = loadBytes(TBinary.bytesOffset(boff), length)
+        v.visitBinary(b)
+      case t: TContainer =>
+        val aoff = loadAddress(off)
+        val length = t.loadLength(this, aoff)
+        v.enterArray(t, length)
+        var i = 0
+        while (i < length) {
+          v.enterElement(i)
+          if (t.isElementDefined(this, aoff, i))
+            visit(t.elementType, t.elementOffset(aoff, length, i), v)
+          else
+            v.visitMissing(t.elementType)
+          i += 1
+        }
+        v.leaveArray()
+      case t: TStruct =>
+        v.enterStruct(t)
+        var i = 0
+        while (i < t.size) {
+          val f = t.fields(i)
+          v.enterField(f)
+          if (t.isFieldDefined(this, off, i))
+            visit(f.typ, t.fieldOffset(off, i), v)
+          else
+            v.visitMissing(f.typ)
+          v.leaveField()
+          i += 1
+        }
+        v.leaveStruct()
+      case t: ComplexType =>
+        visit(t.representation, off, v)
+    }
+  }
+
+  def pretty(t: Type, off: Long): String = {
+    val v = new PrettyVisitor()
+    visit(t, off, v)
+    v.result()
+  }
 }
 
-case class RegionValue(region: MemoryBuffer, offset: Long)
+trait ValueVisitor {
```

Comment:
A comment: I'm inclined to add `result(): T` and make this generic, but it's not relevant to our current use-case.

Reply:
I have imagined use cases that won't return a result.  For example, the encoder should take a region to encode into (which might already have stuff in it) and the value to write, but not return anything.